libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineRoute.kt

```diff
@@ -0,0 +1,238 @@
+            return Builder(routeUrl)
+        }
+    }
+
+    /**
+     * Builds a URL string for offline.
+     *
+     * @return the offline url string
+     */
+    fun buildUrl(): String {
+        return buildOfflineUrl(routeUrl.getRequest())
+    }
+
+    private fun checkWaypointTypes(waypointTypes: List<OfflineCriteria.WaypointType?>?): String? {
+        return if (waypointTypes.isNullOrEmpty()) {
+            null
+        } else {
+            formatWaypointTypes(waypointTypes)
+                ?: throw IllegalStateException("All waypoint types values must be one of break, through or null")
+        }
+    }
+
+    private fun formatWaypointTypes(waypointTypesToFormat: List<OfflineCriteria.WaypointType?>): String? {
+        val waypointTypes = waypointTypesToFormat.map { it?.type ?: "" }.toTypedArray()
+        return waypointTypes.joinTo(StringBuilder(), ";").toString()
+    }
+
+    private fun buildOfflineUrl(url: Uri): String {
+        val offlineUrlBuilder = url.buildUpon()
+        ifNonNull(bicycleType) {
+            offlineUrlBuilder.appendQueryParameter(BICYCLE_TYPE_QUERY_PARAMETER, it)
+        }
+
+        ifNonNull(cyclingSpeed) {
+            offlineUrlBuilder.appendQueryParameter(CYCLING_SPEED_QUERY_PARAMETER, it.toString())
+        }
+
+        ifNonNull(cyclewayBias) {
+            offlineUrlBuilder.appendQueryParameter(CYCLEWAY_BIAS_QUERY_PARAMETER, it.toString())
+        }
+
+        ifNonNull(hillBias) {
+            offlineUrlBuilder.appendQueryParameter(HILL_BIAS_QUERY_PARAMETER, it.toString())
+        }
+
+        ifNonNull(ferryBias) {
+            offlineUrlBuilder.appendQueryParameter(FERRY_BIAS_QUERY_PARAMETER, it.toString())
+        }
+
+        ifNonNull(roughSurfaceBias) {
+            offlineUrlBuilder.appendQueryParameter(
+                ROUGH_SURFACE_BIAS_QUERY_PARAMETER,
+                it.toString()
+            )
+        }
+
+        ifNonNull(waypointTypes) {
+            offlineUrlBuilder.appendQueryParameter(WAYPOINT_TYPES_QUERY_PARAMETER, it)
+        }
+        return offlineUrlBuilder.build().toString()
+    }
+
+    class Builder internal constructor(private val routeUrl: RouteUrl) {
+        private var bicycleType: OfflineCriteria.BicycleType? = null
+        private var cyclingSpeed: Float? = null
+        private var cyclewayBias: Float? = null
+        private var hillBias: Float? = null
+        private var ferryBias: Float? = null
+        private var roughSurfaceBias: Float? = null
+        private var waypointTypes: List<OfflineCriteria.WaypointType?>? = null
+
+        /**
+         * The type of bicycle, either <tt>Road</tt>, <tt>Hybrid</tt>, <tt>City</tt>, <tt>Cross</tt>, <tt>Mountain</tt>.
+         * The default type is <tt>Hybrid</tt>.
+         *
+         * @param bicycleType the type of bicycle
+         * @return this builder for chaining options together
+         */
+        fun bicycleType(bicycleType: OfflineCriteria.BicycleType?): Builder {
```

Comment:
All these functions `bicycleType`, `cyclingSpeed`, etc. should be `internal`. See comment above.

Reply:
make `OfflineRoute` internal